pgxn/neon/pagestore_smgr.c

```diff
@@ -2403,19 +2402,10 @@ neon_get_request_lsns(NRelFileInfo rinfo, ForkNumber forknum, BlockNumber blkno,
 			 * primary, but for the primary we can avoid it by always
 			 * requesting the latest page, by setting request LSN to
 			 * UINT64_MAX.
-			 *
-			 * Remember the current LSN, however, so that we can later
-			 * correctly determine if the response to the request is still
-			 * valid. The most up-to-date LSN we could use for that purpose
-			 * would be the current insert LSN, but to avoid the overhead of
-			 * looking it up, use 'flushlsn' instead. This relies on the
-			 * assumption that if the page was modified since the last WAL
-			 * flush, it should still be in the buffer cache, and we
-			 * wouldn't be requesting it.
 			 */
 			result->request_lsn = UINT64_MAX;
 			result->not_modified_since = last_written_lsn;
-			result->effective_request_lsn = flushlsn;
+			result->effective_request_lsn = last_written_lsn;
```

Comment:
> It should not cause some degrade of performance because we store LwLSN when it was not found in LwLSN hash, so if page is not changed till prefetch response is arrived, then LwLSN should not be changed.

I suppose. I'd rather not rely on that if possible, though. The entry might be found in the LwLSN hash, and evicted before the response arrives, for example. Not sure how common that is with the size of the cache. Might there be some pathological pattern with a SeqScan evicting everything from the LwLSN cache just ahead of prefetch requests or something?

How about calling GetFlushRecPtr() before GetLastWrittenLSNv()?

Reply:
>  I'd rather not rely on that if possible, though
We are not relying on it (I sense that correctness can not violated if LwLSN was evicted from cache).

> How about calling GetFlushRecPtr() before GetLastWrittenLSNv()?
It was my first attempt to fix this issue:)
it doesn't help. It takes me some time to understand why.
Change of the pages can be wallogged, but its LwLSN is not updated. It will be updated only when page is thrown away from shared buffers and written to SMGR.  So the same scenario in the description of the ticket still can happen.

So the only safe upper boundary for `effective_lsn` is such LSN that there are not WAL records modifying this page in interval `[not_modified_since,effective_lsn]`. But there seems to be no simple and efficient way to get such LSN.

Reply:
I see :-(. You'd need to call GetFlushRecPtr() before checking if the page is in the buffer cache. But you can't do that within smgrprefetch(), would need to modify the caller.

Could we use the continue using `flush_lsn` for smgrread() calls, when the buffer is already mapped in the buffer cache, and only use `last_written_lsn` for prefetch requests? Or does it make no difference?

Reply:
Yes, `effective_lsn` makes sense only for prefetch requests - for validation of returned response.